### PR TITLE
Implementing SPARQL query externalization for "Project" and "Clinical Trial" aspects (issue #1282)

### PR DESCRIPTION
**Pull request by ammar257ammar**
As assigned by @Daniel-Mietchen  (issue #1282  )
 
The implementation followed @fnielsen example #1267 

Tox check passed successfully

### CODE REVIEW

File: scholia/app/templates/clinical_trial.html [hunk 1]
```diff
@@ -157,12 +20,12 @@ <h1 id="h1">Clinical trial</h1>
 
 <h2>Data</h2>
```
Comment (egonw) [inline scholia/app/templates/clinical_trial.html:21]: @ammar257ammar, the original pattern used by Finn also has IDs on the header elements, with the ID name like `foo-header` for `foo-table`. So, here you'd have `id="data.header"`.

**Review by egonw — CHANGES_REQUESTED** [review]

File: scholia/app/templates/clinical_trial.html (hunk 1 above)
Comment (fnielsen) [reply scholia/app/templates/clinical_trial.html:21]: Wouldn't it be `id="data-header"`?

File: scholia/app/templates/project.html [hunk 2]
```diff
@@ -118,20 +28,20 @@ <h1 id="h1">Project</h1>
 <h2 id="Project Partners">Project Partners</h2>
 
 <div class="embed-responsive embed-responsive-4by3">
-  <iframe class="embed-responsive-item"  src="https://query.wikidata.org/embed.html#%23defaultView%3AMap%0ASELECT%20%3Fpartner%20%3FpartnerLabel%20%3Flocation%0AWITH%20%7B%0A%20%20SELECT%20DISTINCT%20%3Fpartner%20%3Flocation%20WHERE%20%7B%0A%20%20%20%20wd%3A{{ q }}%20%28%5Ewdt%3AP1344%20%7C%20wdt%3AP710%29%2a%20%2F%20%28wdt%3AP527%20%7C%20wdt%3AP710%20%7C%20wdt%3AP1416%29%20%3Fpartner%20.%0A%20%20%20%20%3Fpartner%20wdt%3AP159%2a%20%2F%20wdt%3AP625%20%3Flocation%20.%0A%20%20%7D%0A%7D%20AS%20%25result%0AWHERE%20%7B%0A%20%20INCLUDE%20%25result%0A%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22%5BAUTO_LANGUAGE%5D%2Cen%22.%20%7D%0A%7D%0A"></iframe>
+  <iframe class="embed-responsive-item" id="partners-iframe"></iframe>
```
Comment (fnielsen) [inline scholia/app/templates/project.html:31]: There should be a correspondence between h2 id and iframe id.

File: scholia/app/templates/project.html [hunk 3]
```diff
@@ -118,20 +28,20 @@ <h1 id="h1">Project</h1>
...
 </div>
 
 
 <h2>Timeline</h2>
```
Comment (fnielsen) [inline scholia/app/templates/project.html:35]: `timeline-header`

File: scholia/app/templates/project.html [hunk 4]
```diff
@@ -118,20 +28,20 @@ <h1 id="h1">Project</h1>
...
+  <iframe class="embed-responsive-item" id="timeline-iframe"></iframe>
 </div>
 
 
 <h2 id="Recently published works about or funded by the project">Recently published works about or funded by the project</h2>
```
Comment (fnielsen) [inline scholia/app/templates/project.html:42]: `recent-articles-header`

File: scholia/app/templates/project.html [hunk 5]
```diff
@@ -118,20 +28,20 @@ <h1 id="h1">Project</h1>
...
 
-<table class="table table-hover" id="recentArticles"></table>
+<table class="table table-hover" id="recent-articles-table"></table>
 
 
 <h2>Topic scores</h2>
```
Comment (fnielsen) [inline scholia/app/templates/project.html:47]: `id="topic-scores"`

File: scholia/app/templates/project.html [hunk 6]
```diff
@@ -140,19 +50,19 @@ <h2>Topic scores</h2>
...
+  <iframe class="embed-responsive-item" id="topic-scores-iframe"></iframe>
 </div>
 
 <h2 id="Authors">Authors</h3>
 
 <h3 id="Prolific authors">Prolific authors</h3>
```
Comment (fnielsen) [inline scholia/app/templates/project.html:58]: `id="prolific-authors-header"`

File: scholia/app/templates/project.html [hunk 7]
```diff
@@ -140,19 +50,19 @@ <h2>Topic scores</h2>
...
 
-<table class="table table-hover" id="prolific-authors"></table>
+<table class="table table-hover" id="prolific-authors-table"></table>
 
 <h2 id="Publications per year">Publications per year</h2>
```
Comment (fnielsen) [inline scholia/app/templates/project.html:62]: `id="publications-per-year"`

File: scholia/app/templates/clinical_trial.html (hunk 1 above)
Comment (fnielsen) [reply scholia/app/templates/clinical_trial.html:21]: @ammar257ammar I have added a few other places where the old "ids" are wrong.

File: scholia/app/templates/project.html (hunk 5 above)
Comment (egonw) [reply scholia/app/templates/project.html:47]: id="topic-scores-header" :)

**Comment by fnielsen** [comment]
@ammar257ammar @egonw one important issue is that I originally thought that the aspects file names should be with dashes rather than underscore, e.g., like "clinical-trial_ prolific-authors".

**Comment by ammar257ammar** [comment]
Thank you @egonw and @fnielsen  for the feedback. I am fixing it now.
 @fnielsen, by changing the aspect file names (e.g., like "clinical-trial_ prolific-authors"), does that include the HTML templates or just the SPARQL queries, and if so do we need to change the line in the code:
 {% set aspect = "clinical_trial" %}
 To
 {% set aspect = "clinical-trial" %}

Please confirm if this is true or not.
Thanks

**Comment by ammar257ammar** [comment]
Implementation is updated with your feedback notes, except for the aspect underscore replacement (waiting your confirmation in the previous comment)
Thanks

**Comment by fnielsen** [comment]
Yes, "clinical-trial" is the way I suggest. Sorry for not specifying that before.

**Comment by fnielsen** [comment]
The specification of the display type is missing for the iframe panels in both clinical-trial and project aspects.

**Comment by fnielsen** [comment]
So I suppose we are ready to go now?

**Comment by ammar257ammar** [comment]
Yes, we are ready. I replaced the underscore with dash in the aspect name and in the call like this:
{% set aspect = "clinical-trial" %}
Check it if it is true, then you can merge
Thanks

File: scholia/app/templates/project.html (hunk 2 above)
Comment (egonw) [reply scholia/app/templates/project.html:31]: @ammar257ammar, this still seems to need fixing.

**Review by egonw — CHANGES_REQUESTED** [review]
@ammar257ammar, one fix left, I think.

**Comment by ammar257ammar** [comment]
@egonw this issue is already fixed in this commit
https://github.com/ammar257ammar/scholia/commit/7f0a7d0ee70db13793cf0f1d0c10d19b70325024

**Comment by egonw** [comment]
> @egonw this issue is already fixed in this commit
> [ammar257ammar@7f0a7d0](https://github.com/ammar257ammar/scholia/commit/7f0a7d0ee70db13793cf0f1d0c10d19b70325024)

Sorry, I do not see that. the header is `project-partners-header` while the iframe is `partners-iframe`. That doesn't match. It should be either `partners-header` with `partners-iframe` OR `project-partners-header` with `project-partners-iframe`.

**Comment by ammar257ammar** [comment]
Oh, I got you. My mistake.
Here is the fix
https://github.com/fnielsen/scholia/pull/1307/commits/25560df9eda678df0e06bf1e3308b7034edd0030

**Review by egonw — APPROVED** [review]

**Comment by fnielsen** [comment]
I get 

```
jinja2.exceptions.TemplateNotFound

TemplateNotFound: clinical_trial_empty.html
```

**Comment by egonw** [comment]
> clinical-trial

**Comment by egonw** [comment]
> Yes, "clinical-trial" is the way I suggest. Sorry for not specifying that before.

This is going to be a problem. The underscore is there, because it has to be a underscore in the method name. It must be "clinical_trial"

**Pull request merged**